### DISCUSSION TITLE
Flows - derived data, simplified.

### DISCUSSION THREAD

**Discussion by kimo-k** [day8/re-frame#795] [General]
Flows are a proposed feature for the next version of re-frame.

- [Live docs](https://day8.github.io/re-frame/Flows/)
- [Example project](https://github.com/day8/re-frame/tree/master/examples/flow)
- [Slack discussion](https://clojurians.slack.com/archives/C073DKH9P/p1698792674379499)

**Comment by light-matters** [comment]
Is there an ETA for this feature (i.e. for when it will leave alpha)? This is something that I'd be very excited about.

**Comment by kimo-k** [comment]
Hey there. We're hoping to get back to open-source work in the spring. But in general, we prefer to keep things in alpha for a long time. I'm still not sure if flows are right for re-frame. Maybe re-frame.subs.alpha is better? Or maybe alpha subscriptions and flows can be consolidated into a single feature. Any critique you have to share would be very useful.

**Comment by kimo-k** [comment]
Hi @rads,

>>I'm not sure flows can replace subscriptions completely, though.

> My initial reaction is that flows replacing subscriptions would be a good thing since that means fewer concepts to deal with in the long run. I'm probably missing the full context though. Would you mind explaining a bit more why you feel this way? Is it a technical limitation? Different use cases? Backwards compatibility?

Flows store their output in app-db. I agree that this lowers some conceptual overhead. It also makes it very straightforward for an event handler access a flow's output value. By default, that value is eagerly computed, and cached forever. You can pause that computation using the `:live?` predicate, or imperatively create/destroy the flow using `:reg-flow` and `:clear-flow`. 

On the other hand, flows change re-frame's aesthetic. Re-frame is arguably a form of "guard rails", promising a unified way to manage state. But, flows might enable us to make bigger messes.

Subscriptions do feel well "separated" from the "essential state" ([Mosely & Marks, 7.3.2](https://github.com/papers-we-love/papers-we-love/blob/main/design/out-of-the-tar-pit.pdf)), i.e., app-db. For instance, a dev wouldn't consider swapping a subscription's cached value as a way to fix a bug, or build a feature.

Flows don't have that separation. By registering a graph of flows, have we diluted our essential state with accidental state & control? It might be up to us programmers to resist ambiguity - for instance, putting flow outputs under a dedicated `:flow-outputs` path. Maybe re-frame should print a warning when the subtree located at a flow's output path is changed by something other than a flow.

I have colleagues who write all their own subscriptions and events. We don't have to understand each other's code. Instead, the structure of app-db is our common reference. With flows, how will they know what parts of app-db they can change (essential state), and what parts they should leave alone (derived state)?

That said, I am still enjoying flows. They feel like backwards events. 
- With events, when the user needs something, that means app-db should change. 
- With flows, when app-db changes, that means the user needs something. 

This makes for some clean solutions.

I currently maintain some complicated input forms. When their values change, we need to issue an api request. When api results arrive, we need to post-process and draw them. Except, there are so many details...  
- Some changes don't need an api call, just different post-processing. 
- Some changes only need to be redrawn. 
- Some changes should be batched, a.k.a., debounced. 
- Some changes shouldn't take any effect until the user actually goes to the tab where they can see the results. 

I use flows to express all those little details, purely in terms of app-db. Then, I can keep the form fields extremely basic. They just put their new values into app-db, totally naive of any API or view requirements.
